cosmwasm/packages/sgx-vm/src/wasmi/wrapper.rs

```diff
@@ -4,22 +4,47 @@ use std::ffi::c_void;
 use std::marker::PhantomData;
 use std::mem::MaybeUninit;
 
+use crate::enclave::get_query_token;
 use crate::errors::{EnclaveError, VmResult};
-use crate::{Querier, Storage};
+use crate::{Querier, Storage, VmError};
 
-use enclave_ffi_types::{Ctx, HandleResult, InitResult, QueryResult};
+use enclave_ffi_types::{Ctx, EnclaveBuffer, HandleResult, InitResult, QueryResult};
 
-use sgx_types::sgx_status_t;
+use sgx_types::{sgx_status_t, SgxResult};
+use sgx_urts::SgxEnclave;
 
 use log::*;
+use serde::Deserialize;
 
 use super::exports::FullContext;
 use super::imports;
 use super::results::{
     handle_result_to_vm_result, init_result_to_vm_result, query_result_to_vm_result, HandleSuccess,
     InitSuccess, QuerySuccess,
 };
-use crate::enclave::EnclaveGuard;
+
+/// This is a safe wrapper for allocating buffers inside the enclave.
+pub(super) fn allocate_enclave_buffer(buffer: &[u8]) -> SgxResult<EnclaveBuffer> {
+    let ptr = buffer.as_ptr();
+    let len = buffer.len();
+    let mut enclave_buffer = EnclaveBuffer::default();
+
+    let enclave_id = crate::enclave::get_enclave()
+        .expect("If we got here, surely the enclave has been loaded")
```

Comment:
I think this message should be changed, because now the enclave in not Mutex protected?